service/history/ndc/mutable_state_mapper.go

```diff
@@ -0,0 +1,182 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ndc
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+
+	"go.temporal.io/server/common/log/tag"
+	serviceerrors "go.temporal.io/server/common/serviceerror"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+type (
+	MutableStateMapper[Input any, Output any] func(
+		ctx context.Context,
+		wfContext workflow.Context,
+		mutableState workflow.MutableState,
+		input Input,
+	) (workflow.MutableState, Output, error)
+
+	MutableStateMapperImpl struct {
+		shardContext          shard.Context
+		newBufferEventFlusher bufferEventFlusherProvider
+		newBranchMgr          branchMgrProvider
+		newConflictResolver   conflictResolverProvider
+		newStateBuilder       stateBuilderProvider
+	}
+
+	PrepareHistoryBranchOut struct {
+		DoContinue  bool
+		BranchIndex int32
+	}
+
+	PrepareMutableStateIn struct {
+		replicationTask
+		BranchIndex int32
+	}
+)
+
+var _ MutableStateMapper[replicationTask, struct{}] = (*MutableStateMapperImpl)(nil).FlushBufferEvents
+var _ MutableStateMapper[replicationTask, PrepareHistoryBranchOut] = (*MutableStateMapperImpl)(nil).PrepareHistoryBranch
+var _ MutableStateMapper[PrepareMutableStateIn, bool] = (*MutableStateMapperImpl)(nil).PrepareMutableState
+var _ MutableStateMapper[replicationTask, workflow.MutableState] = (*MutableStateMapperImpl)(nil).ApplyEvents
+
+func NewMutableStateMapping(
+	shardContext shard.Context,
+	newBufferEventFlusher bufferEventFlusherProvider,
+	newBranchMgr branchMgrProvider,
+	newConflictResolver conflictResolverProvider,
+	newStateBuilder stateBuilderProvider,
+) *MutableStateMapperImpl {
+	return &MutableStateMapperImpl{
+		shardContext:          shardContext,
+		newBufferEventFlusher: newBufferEventFlusher,
+		newBranchMgr:          newBranchMgr,
+		newConflictResolver:   newConflictResolver,
+		newStateBuilder:       newStateBuilder,
+	}
+}
+
+func (m *MutableStateMapperImpl) FlushBufferEvents(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task replicationTask,
+) (workflow.MutableState, struct{}, error) {
+	flusher := m.newBufferEventFlusher(wfContext, mutableState, task.getLogger())
+	_, mutableState, err := flusher.flush(ctx)
```

Comment:
do you need to return the context?

Reply:
context is not expected to change, but mutable state can, e.g. ms rebuild due to conflict